grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java

```diff
@@ -199,21 +255,24 @@ public void start(Listener<O> responseListener, Metadata metadata) {
         requireNonNull(responseListener, "responseListener");
         requireNonNull(metadata, "metadata");
 
+        // The listener is set by the callers thread. The safe-publishing (aka visibility)
+        // to event-loop is guaranteed via semantics of EventLoop.execute(...) call.
+        listener = new OffloadingListener(responseListener);
```

Comment:
We probably treat this change as optimization and avoid allocating `OffloadingListener` but we'd have to get back to managing context push manually (which is now done via `ContextAwareExecutor(directExecutor)`) in case when listener isn't `OffloadingListener`. This adds complexity and I'm not (yet) sure it's worth optimizing. If we want optimization - I'd suggest getting rid of `OffloadingListener` and inline calls to `contextAwareExecutor` like it was before - the intent was clear and it didn't cost allocations.

Reply:
I didn't know that a request context is not propagated to the `Listener` that sounds a bug.

What do you think of directly pushing the context to the current thread if `callOptions.getExecutor()` is `null` although it is more verbose than your version? We provide a range of wrappers to propagate contexts but internally use `ctx.push()` with `try-with-resources` to minimize allocations.

```java
class OffloadingListener {

    @Override
    public void onHeaders(Metadata metadata) {
        if (contextAwareExecutor == null) {
            try (AutoCloseable ignored = ctx.push()) {
                delegate.onHeaders(metadata);
            }
        } else {
            contextAwareExecutor.execute(() -> {
                try {
                    delegate.onHeaders(metadata);
                } catch (Throwable t) {
                    closeWhenListenerThrows(t);
                }
            });
        }
    }

}
```

That said, if it looks overkill, I'm okay with the current implementation.

Reply:
By the way, it looks useful to specify an exception handler to `contextAwareExecutor` when it is created.

```java
ctx.makeContextAware(MoreExecutors.newSequentialExecutor(
    callOptions.getExecutor()), cause -> closeWhenListenerThrows(cause));
```

Reply:
Hmm, I don't see this `makeContextAware` method that takes exception handler in the API. Is there a link?

I think it's a good idea to try to optimize it. I agree we should avoid excessive allocations while it's reasonable and doesn't hurt redability. My proposal is to keep relying on `MoreExecutors.directExecutor()` to unify handling transport closures like it's done [here](https://github.com/line/armeria/pull/4851/files#diff-39112577712e2be6800f84834abc33a690eed333cd96581b777e1e1cd1ccdadaR499).

What we can and probably should do to save on allocations is

1) Get rid of OffloadingListener and keep inlining closeWhenListenerThrows()
2) Get rid of ContextAwareExecutor and keep inlining ctx.push()

These two give us enough savings (at least less 5 runnables allocated on a typical unary RPC call) to keep directExecutor() around for simplicity reasons. WDYT?

Reply:
> Hmm, I don't see this makeContextAware method that takes exception handler in the API. Is there a link?

Sorry, we don't have such an API. It was a proposal. I created an issue for the idea. #4879

> These two give us enough savings (at least less 5 runnables allocated on a typical unary RPC call) to keep directExecutor() around for simplicity reasons. WDYT?

Sound good. If we can reduce the 5 allocations with some effort, I think it will be well worth it.

Reply:
@ikhoon Updated the PR to optimize for lower allocations. PTAL